pkg/router/controller/status.go

```diff
@@ -196,14 +196,15 @@ func recordIngressCondition(route *routev1.Route, name, hostName string, conditi
 			return false, false, time.Time{}, existing, existing
 		}
 
-		// preserve a copy of the original ingress without conditions
-		original := *existing
-		original.Conditions = nil
+		// preserve a deep copy of the original ingress
+		original := existing.DeepCopy()
```

Comment:
> This seems like a pessimization.

You have shared ownership of the underlying array tied to the Conditions slice. If `existing` goes on to modify (i.e., reallocate) Conditions then the shallow copy is OK. But if no reallocation takes place then mutations to `existing.Conditions[x].Foo` will reflect in `original` too.

Reply:
Does this help explain my concern? 

```go
package main_test

import (
	"testing"
	"time"
)

var unixEpochTime = time.Unix(0, 0)

const epochMsg string = "Epoch"

type Condition struct {
	Type               string
	Status             string
	LastTransitionTime *time.Time
	Message            string
}

type Ingress struct {
	Host       string
	Conditions []Condition
}

type mutateConditionOp int

const (
	MutateExistingCondition mutateConditionOp = iota
	AddNewCondition
)

func findIngress(ingress *Ingress) *Ingress {
	// Return a shallow copy.
	original := *ingress
	return &original
}

func doStuffWithIngress(ingress *Ingress, op mutateConditionOp) {
	switch op {
	case MutateExistingCondition:
		for i := range ingress.Conditions {
			newTime := time.Now()
			ingress.Conditions[i].LastTransitionTime = &newTime
			ingress.Conditions[i].Message = "Mutated condition"
		}
	case AddNewCondition:
		// newTime := time.Now()
		// ingress.Conditions[0].LastTransitionTime = &newTime
		// ingress.Conditions[0].Message = "Mutated condition"
		//    or ...
		// ingress.Conditions = nil
		for i := 0; i < 9; i++ {
			newTime := time.Now()
			newCondition := Condition{
				Type: "Info", Status: "True", LastTransitionTime: &newTime, Message: "Additional condition",
			}
			ingress.Conditions = append(ingress.Conditions, newCondition)
		}
	}
}

// Test to demonstrate the surprise mutation due to shallow copying.
func TestUnexpectedMutation(t *testing.T) {
	ingress := Ingress{
		Host: "example.com",
		Conditions: []Condition{
			{Type: "Ready", Status: "True", LastTransitionTime: &unixEpochTime, Message: epochMsg},
		},
	}

	original := findIngress(&ingress)

	// Another part of the system unknowing to all and sundry
	// mutates an condition.
	doStuffWithIngress(&ingress, MutateExistingCondition)

	if !original.Conditions[0].LastTransitionTime.Equal(unixEpochTime) {
		t.Errorf("Expected original.Conditions[0].LastTransitionTime to be %q, got %q, so something somehere mutated the shallow copy", unixEpochTime, original.Conditions[0].LastTransitionTime)
	}

	if original.Conditions[0].Message != epochMsg {
		t.Errorf("Expected original.Conditionss[0].Message to be %q, got %q, so something somewhere mutated the shallow copy", epochMsg, original.Conditions[0].Message)
	}
}

// Test to demonstrate no mutation due to slice realloc.
func TestNoMutationBecauseOfSliceReallocation(t *testing.T) {
	ingress := Ingress{
		Host: "example.com",
		Conditions: []Condition{
			{Type: "Ready", Status: "True", LastTransitionTime: &unixEpochTime, Message: epochMsg},
		},
	}

	original := findIngress(&ingress)

	// Another part of the system unknowing to all and sundry
	// mutates the number of conditions.
	doStuffWithIngress(&ingress, AddNewCondition)

	// Verify that the "original" object was NOT mutated because
	// the underlying array backing ingress.Conditions got
	// reallocated.
	if e, l := 1, len(original.Conditions); e != l {
		t.Errorf("Expected `len(original.Conditions) %d, got %d", e, l)
	}

	if e, l := 1, cap(original.Conditions); e != l {
		t.Errorf("Expected `cap(original.Conditions) %d, got %d", e, l)
	}

	if !original.Conditions[0].LastTransitionTime.Equal(unixEpochTime) {
		t.Errorf("Expected original.Conditions[0].LastTransitionTime to remain unchanged, got %q, so something mutated the shallow copy", *original.Conditions[0].LastTransitionTime)
	}

	if original.Conditions[0].Message != epochMsg {
		t.Errorf("Expected original.Conditions[0].Message to remain %q, got %q, so something mutated the shallow copy", original.Conditions[0].Message, epochMsg)
	}

	if e, l := 10, len(ingress.Conditions); l != e {
		t.Errorf("Expected len(ingress.Conditions) = %v, got %v", e, l)
	}
}
```